// ngx-barcode-scanner/esm2020/lib/barcode-scanner-livestream/barcode-scanner-livestream.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BarcodeScannerLivestreamComponent } from './barcode-scanner-livestream.component';
import * as i0 from "@angular/core";
export class BarcodeScannerLivestreamModule {
}
BarcodeScannerLivestreamModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BarcodeScannerLivestreamModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamModule, declarations: [BarcodeScannerLivestreamComponent], imports: [CommonModule], exports: [BarcodeScannerLivestreamComponent] });
BarcodeScannerLivestreamModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        BarcodeScannerLivestreamComponent
                    ],
                    exports: [
                        BarcodeScannerLivestreamComponent
                    ]
                }]
        }] });
