// @developer-partners/ngx-modal-dialog/esm2020/lib/modal.service.mjs
import { Injectable } from "@angular/core";
import { ModalReference } from "./modal-reference";
import { ModalComponent } from "./modal.component";
import * as i0 from "@angular/core";
export class ModalService {
    constructor(componentFactoryResolver, applicaionRef, injector) {
        this._componentFactoryResolver = componentFactoryResolver;
        this._applicaionRef = applicaionRef;
        this._injector = injector;
    }
    createModalComponent(dependecies) {
        let factory = this._componentFactoryResolver.resolveComponentFactory(ModalComponent);
        return factory.create(new ModalInjector(this._injector, dependecies));
    }
    createModalRefernce(map, config) {
        let modalReference = new InternalModalRef(config);
        // Create with internal implementation, but inject the public version.
        map.set(ModalReference, modalReference);
        modalReference.componentRef = this.createModalComponent(map);
        modalReference
            .result()
            .subscribe({
            complete: () => this.removeModal(modalReference)
        });
        return modalReference;
    }
    show(componentType, config) {
        let map = new WeakMap();
        let modalReference = this.createModalRefernce(map, config);
        modalReference.componentRef.instance.contentComponentType = componentType;
        this._applicaionRef.attachView(modalReference.componentRef.hostView);
        document.body.appendChild(modalReference.getHtmlElement());
        // To trigger the animation effect. The modal is added to document at this point, but it's invisible.
        // We animate the showing part.
        window.setTimeout(() => {
            modalReference.getHtmlElement().firstElementChild.classList.add('dp-visible');
        });
        modalReference.componentRef.changeDetectorRef.detectChanges();
        return modalReference;
    }
    removeModal(modalReference) {
        modalReference.getHtmlElement().firstElementChild.classList.remove('dp-visible');
        // Wait until the closing animation is done, then remove the component.
        window.setTimeout(() => {
            this._applicaionRef.detachView(modalReference.componentRef.hostView);
            modalReference.componentRef.destroy();
        }, 400);
    }
}
ModalService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.2", ngImport: i0, type: ModalService, deps: [{ token: i0.ComponentFactoryResolver }, { token: i0.ApplicationRef }, { token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable });
ModalService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.2", ngImport: i0, type: ModalService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.2", ngImport: i0, type: ModalService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.ComponentFactoryResolver }, { type: i0.ApplicationRef }, { type: i0.Injector }]; } });
class ModalInjector {
    constructor(injector, extraDependencies) {
        this._injector = injector;
        this._extraDependecnies = extraDependencies;
    }
    get(token, notFoundValue, flags) {
        let resolved = this._extraDependecnies.get(token);
        if (resolved) {
            return resolved;
        }
        return this._injector.get(token, notFoundValue, flags);
    }
}
class InternalModalRef extends ModalReference {
    constructor(config) {
        super(config);
    }
    getHtmlElement() {
        let viewRef = this.componentRef.hostView;
        return viewRef.rootNodes[0];
    }
}
