// ngx-barcode-scanner/esm2020/lib/barcode-scanner-livestream-overlay/barcode-scanner-livestream-overlay.component.mjs
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BarcodeScannerLivestreamComponent } from '../barcode-scanner-livestream/barcode-scanner-livestream.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../barcode-scanner-livestream/barcode-scanner-livestream.component";
export class BarcodeScannerLivestreamOverlayComponent {
    constructor() {
        this._started = false;
        this.width = '90vw';
        this.maxWidth = '640px';
        this.valueChanges = new EventEmitter();
        this.started = new EventEmitter();
        this._showScanner = false;
    }
    get isStarted() {
        return this._started;
    }
    get showScanner() {
        return this._showScanner;
    }
    ngOnDestroy() {
        this.scanner.stop();
    }
    show() {
        this._showScanner = true;
        this.scanner.start();
    }
    hide() {
        this._showScanner = false;
        this.scanner.stop();
    }
    onStarted(value) {
        this._started = value;
        this.started.next(value);
    }
    onValueChanges(result) {
        this.valueChanges.next(result);
    }
}
BarcodeScannerLivestreamOverlayComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
BarcodeScannerLivestreamOverlayComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.11", type: BarcodeScannerLivestreamOverlayComponent, selector: "barcode-scanner-livestream-overlay", inputs: { type: "type", deviceId: "deviceId", width: "width", maxWidth: "maxWidth", height: "height", maxHeight: "maxHeight", config: "config" }, outputs: { valueChanges: "valueChanges", started: "started" }, viewQueries: [{ propertyName: "scanner", first: true, predicate: BarcodeScannerLivestreamComponent, descendants: true }], ngImport: i0, template: "<div class=\"barcode-scanner-livestream-overlay\" [hidden]=\"!showScanner\">\n  <div\n    class=\"barcode-scanner-livestream-overlay-content\"\n    [style.width]=\"width\"\n    [style.max-width]=\"maxWidth\"\n    [style.height]=\"height\"\n    [style.max-height]=\"maxHeight\"\n  >\n    <div\n      class=\"barcode-scanner-livestream-overlay-close\"\n      *ngIf=\"isStarted\"\n      (click)=\"hide()\"\n    >\n      X\n    </div>\n    <barcode-scanner-livestream\n      [type]=\"type\"\n      [deviceId]=\"deviceId\"\n      [config]=\"config\"\n      (valueChanges)=\"onValueChanges($event)\"\n      (started)=\"onStarted($event)\"\n    ></barcode-scanner-livestream>\n  </div>\n</div>\n", styles: [".barcode-scanner-livestream-overlay{overflow:hidden;position:fixed;inset:0;width:100%;background-color:#0000004d;z-index:1000}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content{top:50%;position:absolute;left:50%;transform:translate(-50%,-50%)}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content .barcode-scanner-livestream-overlay-close{position:absolute;right:0;padding:.5rem;width:1.5rem;height:1.5rem;line-height:1.5rem;text-align:center;background-color:#fff;cursor:pointer;border:2px solid black;font-size:1.3rem;margin:-1rem;border-radius:2rem;z-index:100;box-sizing:content-box}\n"], dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i2.BarcodeScannerLivestreamComponent, selector: "barcode-scanner-livestream", inputs: ["type", "deviceId", "maxWidth", "maxHeight", "config", "errorFilter", "torch"], outputs: ["valueChanges", "started"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayComponent, decorators: [{
            type: Component,
            args: [{ selector: 'barcode-scanner-livestream-overlay', template: "<div class=\"barcode-scanner-livestream-overlay\" [hidden]=\"!showScanner\">\n  <div\n    class=\"barcode-scanner-livestream-overlay-content\"\n    [style.width]=\"width\"\n    [style.max-width]=\"maxWidth\"\n    [style.height]=\"height\"\n    [style.max-height]=\"maxHeight\"\n  >\n    <div\n      class=\"barcode-scanner-livestream-overlay-close\"\n      *ngIf=\"isStarted\"\n      (click)=\"hide()\"\n    >\n      X\n    </div>\n    <barcode-scanner-livestream\n      [type]=\"type\"\n      [deviceId]=\"deviceId\"\n      [config]=\"config\"\n      (valueChanges)=\"onValueChanges($event)\"\n      (started)=\"onStarted($event)\"\n    ></barcode-scanner-livestream>\n  </div>\n</div>\n", styles: [".barcode-scanner-livestream-overlay{overflow:hidden;position:fixed;inset:0;width:100%;background-color:#0000004d;z-index:1000}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content{top:50%;position:absolute;left:50%;transform:translate(-50%,-50%)}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content .barcode-scanner-livestream-overlay-close{position:absolute;right:0;padding:.5rem;width:1.5rem;height:1.5rem;line-height:1.5rem;text-align:center;background-color:#fff;cursor:pointer;border:2px solid black;font-size:1.3rem;margin:-1rem;border-radius:2rem;z-index:100;box-sizing:content-box}\n"] }]
        }], propDecorators: { type: [{
                type: Input
            }], deviceId: [{
                type: Input
            }], width: [{
                type: Input
            }], maxWidth: [{
                type: Input
            }], height: [{
                type: Input
            }], maxHeight: [{
                type: Input
            }], config: [{
                type: Input
            }], valueChanges: [{
                type: Output
            }], started: [{
                type: Output
            }], scanner: [{
                type: ViewChild,
                args: [BarcodeScannerLivestreamComponent]
            }] } });
