// ngx-barcode-scanner/esm2020/lib/barcode-scanner-livestream-overlay/barcode-scanner-livestream-overlay.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BarcodeScannerLivestreamModule } from '../barcode-scanner-livestream/barcode-scanner-livestream.module';
import { BarcodeScannerLivestreamOverlayComponent } from './barcode-scanner-livestream-overlay.component';
import * as i0 from "@angular/core";
export class BarcodeScannerLivestreamOverlayModule {
}
BarcodeScannerLivestreamOverlayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BarcodeScannerLivestreamOverlayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayModule, declarations: [BarcodeScannerLivestreamOverlayComponent], imports: [CommonModule,
        BarcodeScannerLivestreamModule], exports: [BarcodeScannerLivestreamOverlayComponent] });
BarcodeScannerLivestreamOverlayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayModule, imports: [CommonModule,
        BarcodeScannerLivestreamModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.11", ngImport: i0, type: BarcodeScannerLivestreamOverlayModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        BarcodeScannerLivestreamModule
                    ],
                    declarations: [
                        BarcodeScannerLivestreamOverlayComponent
                    ],
                    exports: [
                        BarcodeScannerLivestreamOverlayComponent
                    ]
                }]
        }] });
